fdbserver/MoveKeys.actor.cpp

```diff
@@ -1306,11 +1351,68 @@ ACTOR Future<Void> removeKeysFromFailedServer(Database cx,
 					// Dest is usually empty, but keep this in case there is parallel data movement
 					src.erase(std::remove(src.begin(), src.end(), serverID), src.end());
 					dest.erase(std::remove(dest.begin(), dest.end(), serverID), dest.end());
-					TraceEvent(SevDebug, "FailedServerSetKey", serverID)
-					    .detail("Key", it.key)
-					    .detail("ValueSrc", describe(src))
-					    .detail("ValueDest", describe(dest));
-					tr.set(keyServersKey(it.key), keyServersValue(UIDtoTagMap, src, dest));
+
+					// If the last src server is to be removed, first check if there are dest servers who is
+					// hosting a read-write copy of the keyrange, and move such dest servers to the src list.
+					if (src.empty() && !dest.empty()) {
+						std::vector<UID> newSources =
+						    wait(pickReadWriteServers(&tr, dest, KeyRangeRef(it.key, keyServers[i + 1].key)));
+						for (const UID& id : newSources) {
+							TraceEvent(SevWarn, "FailedServerAdditionalSourceServer", serverID)
+							    .detail("Key", it.key)
+							    .detail("NewSourceServerFromDest", id);
+							dest.erase(std::remove(dest.begin(), dest.end(), id), dest.end());
+							src.push_back(id);
+						}
+					}
+
+					// Move the keyrange to teamForDroppedRange if the src list becomes empty, and also remove the shard
+					// from all dest servers.
+					if (src.empty()) {
+						assert(!teamForDroppedRange.empty());
+
+						// Assign the shard to teamFroDroppedRange in keyServer space.
+						tr.set(keyServersKey(it.key), keyServersValue(UIDtoTagMap, teamForDroppedRange, {}));
+
+						vector<Future<Void>> actors;
+
+						// Unassign the shard from the dest servers.
+						for (const UID& id : dest) {
+							actors.push_back(krmSetRangeCoalescing(&tr,
+							                                       serverKeysPrefixFor(id),
+							                                       KeyRangeRef(it.key, keyServers[i + 1].key),
+							                                       allKeys,
+							                                       serverKeysFalse));
+						}
+						if (!dest.empty()) {
+							TraceEvent(SevWarn, "FailedServerDropRangeFromDest", serverID)
```

Comment:
When I see this trace event, I will have a difficult time to understand what happened unless I am familiar with the code. 
Could we make the event name more intuitive?
For example, would `ShardLossAllReplicasDropShardFromDest` be more intuitive than `FailedServerDropRangeFromDest `?

I feel `ShardLossAllReplicas` is more initiative than `FailedServer`

Reply:
When a shard losses all replicas, I would use SevWarnAlways instead of SevWarn for its related traces.

Reply:
Change applied.